### PR TITLE
var matrix specializations for a-c unary functions in rev

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This is just going through rev and adding var matrix functions for a bunch of functions.

## Tests

`expect_ad_matvar` tests added for all of the functions in here. Though I still need to write tests for `std::vector<var<matrix>>`

## Side Effects

I wrote this by specializing a function for each rev version. But that means we have to write the `scalar_unary` version like

```cpp
template <typename Container,
          require_not_container_st<std::is_arithmetic, Container>* = nullptr,
          require_not_var_matrix_t<Container>* = nullptr>
inline auto fabs(const Container& x) {
  return apply_scalar_unary<fabs_fun, Container>::apply(x);
}
```

Though there may be some other way to do this where instead we specialize `*_fun`

```cpp
struct fabs_fun {
  template <typename T>
  static inline T fun(const T& x) {
    using std::fabs;
    return fabs(x);
  }
};
```

As it's written right now it's nice because it works with `std::vector<var<matrix>>`. eod we need to specialize somewhere so I don't mind having it like we do now. We could also make `var<matrix>` not a container type so then we can remove the `require_not_var_matrix_t<Container>`

## Release notes

Adds arc trig functions for `var<matrix>` along several other unary operators

## Checklist

- [x] Math issue #1805 

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by bbbales2** [comment]
THere are a few functions like this in #2254. I was stuck there on what to do with `std::vector<varmat>` but sounds like you solved that.

Does what you did here also work with `std::vector<std::vector<varmat>>`, etc?

I asked @andrjohns about this in the other thread but maybe you have already solved it (in which case don't worry about it @andrjohns !)

> Though I still need to write tests for std::vector<var<matrix>>

That should be testable now. I had to add these tests for `log_softmax`. I didn't add tests for `std::vector<std::vector<varmat>>` though.

**Comment by SteveBronder** [comment]
Lemme test that. It should at least be close to working. All this does is overload each function so when apply scalar iterates over the elements it should just do the normal thing and differ to the other specializations until it hits varmat and then just do the varmat thing

File: stan/math/prim/fun/acos.hpp
Comment (SteveBronder) [inline stan/math/prim/fun/acos.hpp:49]: @andrjohns I'm having a hard time removing this `require_not_var_matrix_t<Container>` and I'd like to not have to put these everywhere. Right now this function works for any stan scalar or container that does not have an arithmetic scalar type. So it's kind of a mixed bag of types this takes in. Though specializations for each seems like overkill as well so I'm not really sure if we can make this look nicer

**Comment by SteveBronder** [comment]
@bbbales2 fyi I removed var matrix from the container types. It feels like a var matrix is a single unit and we are pretty much always going to have specializations for var matrix. With the container checks we're normally checking whether it's an eigen vector so we can do some specialization or a std vector so we can recursivly go through it (or to map the std vector to an eigen matrix to use the Eigen specialization). So like I think with var matrix it makes sense to not have it count as a container. Does that make sense?

**Comment by rok-cesnovar** [comment]
But if its not a container, is it a scalar? I dont think that make sense.

**Comment by bbbales2** [comment]
Yeah I'm not totally sure what it is either, but I'm also not using it like the other things we call containers. Popping it out seems fine with me.

Seems like whenever we use `var_value`, it is `require_var_vector_t` or `require_var_matrix_t` or `require_rev_matrix_t` or something like that.

I guess we could change `require_container_t` to `require_std_vector_or_eigen_t` lol but that's getting kinda long.

**Comment by bbbales2** [comment]
@SteveBronder this ready to review? I want to just copy whatever you do here for the scalar functions in #2254 so if it's done that's convenient lol.

**Comment by SteveBronder** [comment]
@bbbales2 this should be good to go

File: stan/math/rev/fun/sin.hpp
Comment (bbbales2) [inline stan/math/rev/fun/sin.hpp:54]: sin_vari above used anymore?

File: stan/math/rev/fun/sinh.hpp
Comment (bbbales2) [inline stan/math/rev/fun/sinh.hpp:52]: sinh_vari used anymore?

File: stan/math/rev/fun/tan.hpp
Comment (bbbales2) [inline stan/math/rev/fun/tan.hpp:52]: tan_vari used anymore?

File: stan/math/rev/fun/tanh.hpp
Comment (bbbales2) [inline stan/math/rev/fun/tanh.hpp:52]: tanh vari used anymore?

File: stan/math/rev/functor/apply_scalar_unary.hpp
Comment (bbbales2) [inline stan/math/rev/functor/apply_scalar_unary.hpp:37]: Not quite var

File: test/unit/math/mix/fun/acos_test.cpp
Comment (bbbales2) [inline test/unit/math/mix/fun/acos_test.cpp:35]: The easier way to build these is with initializer lists. Something like:

```
std::vector<Eigen::MatrixXd> A_vec = { A, A, A };
```

File: test/unit/math/mix/fun/atan2_test.cpp
Comment (bbbales2) [inline test/unit/math/mix/fun/atan2_test.cpp:19]: Now iz the time

File: test/unit/math/mix/fun/asinh_test.cpp
Comment (bbbales2) [inline test/unit/math/mix/fun/asinh_test.cpp:26]: I think all these tests should be for varmat with vector, row vector and Matrix vector arguments.

How about:

```cpp
expect_ad_common_vector_matvar(f);
```

And it does:
```cpp
expect_ad_vector_matvar(f, stan::math::to_vector(stan::test::internal::common_args()));
  // not sure if `common_args` is std::vector or not
```

And that function is something like:

```cpp
expect_ad_vector_matvar(F& f, const auto& x) {
  Eigen::VectorXd v = x;
  Eigen::RowVectorXd r = x.transpose();
  Eigen::MatrixXd m(x.size(), 2);
  com_args_m.block(0, 0, com_args_v.size(), 1) = x;
  com_args_m.block(0, 1, com_args_v.size(), 1) = x.reverse(); // reverse just to mix stuff up
  
  expect_ad_matvar(f, v);
  expect_ad_matvar(f, r);
  expect_ad_matvar(f, m);

  std::vector<Eigen::VectorXd> vv = { v, v };
  std::vector<Eigen::RowVectorXd> rv = { r, r };
  std::vector<Eigen::MatrixXd> mv = { m, m };

  expect_ad_matvar(f, vv);
  expect_ad_matvar(f, rv);
  expect_ad_matvar(f, mv);
}
```

Also does `std::vector<std::vector<...>>` work? Don't think it needs tested I'm just curious.

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Review

File: test/unit/math/mix/fun/asinh_test.cpp
Comment (SteveBronder) [reply test/unit/math/mix/fun/asinh_test.cpp:26]: Yeah I think it makes sense to test matrix vector row vector at least to make sure they compile. I'll write something like this up

File: stan/math/prim/fun/acosh.hpp
Comment (SteveBronder) [reply stan/math/prim/fun/acosh.hpp:86]: Actually this can just be `template <typename T, require_container_t<T>* = nullptr>`

Comment (bbbales2) [reply stan/math/prim/fun/acosh.hpp:86]: There were some functions here that didn't have `apply_vector_unary`/Container implementations. I'm assuming it's because the Eigen array functions didn't exist for them to be written (`cbrt`, this, and others). I don't think cleanup is what we're going for in this pull really, so it's fine to leave it.

Comment (bbbales2) [reply stan/math/prim/fun/acosh.hpp:86]: (But if you want to clean things, feel free, happy to review)

Comment (SteveBronder) [reply stan/math/prim/fun/acosh.hpp:86]: Ups actually ignore that, this needs to be able to take in scalar types so that's why we can't use `require_constainer_st`

File: stan/math/prim/functor/apply_vector_unary.hpp
Comment (SteveBronder) [reply stan/math/prim/functor/apply_vector_unary.hpp:166]: This is the only place I think we need this right now. I'm going to make a sweep next week and see if we can clean up some of the `requires` and I can add these kind of things then. Plus this is only used here for now so once we need it in more places we can make it a normal require

File: stan/math/rev/fun/cbrt.hpp
Comment (SteveBronder) [reply stan/math/rev/fun/cbrt.hpp:54]: `.array()` does not have a `.noalias()`

File: stan/math/rev/fun/sin.hpp
Comment (SteveBronder) [reply stan/math/rev/fun/sin.hpp:54]: Nope!

File: test/unit/math/mix/fun/atan2_test.cpp
Comment (SteveBronder) [reply test/unit/math/mix/fun/atan2_test.cpp:19]: This is only for unary stuff so another time

File: stan/math/rev/fun/cbrt.hpp
Comment (t4c1) [reply stan/math/rev/fun/cbrt.hpp]: Also noalias will only speed things up if the outermost operation on the right hand side is a matrix product.

**Review by bbbales2 — COMMENTED** [review]
I changed the fabs/abs tests to work like the rest of them, and also a couple things with docs.

Let me know if this works for you and I'll merge when we get greenlight.

**Comment by SteveBronder** [comment]
Ty! Yes I just went through and removed the `.noalias()` call where we didn't need them so should be cool and good

**Review by bbbales2 — APPROVED** [review]

**Pull request merged**
